plasmapy/particles/decorators.py

```diff
@@ -26,6 +27,19 @@
 from plasmapy.particles.particle_collections import ParticleList, ParticleListLike
 from plasmapy.utils.exceptions import PlasmaPyDeprecationWarning
 
+
+class _CallableDataDict(TypedDict, total=False):
+    allow_custom_particles: bool
+    allow_particle_lists: bool
+    annotations: dict[str, Any]
+    any_of: Optional[Union[str, Iterable[str]]]
+    callable_: Callable[..., Any]
+    exclude: Optional[Union[str, Iterable[str]]]
+    parameters_to_process: list[str]
+    require: Optional[Union[str, Iterable[str]]]
+    signature: inspect.Signature
```

Comment:
An alternative to static type checking for this would be using [pydantic](https://docs.pydantic.dev/latest/).  But that's for another day.

Reply:
Using `TypedDict` here lets us define the types we expect for different keys.